README.md
![Differify](assets/logo.svg)

## One of the Fastest **deep** object/array diff

- Benchmarks with other popular packages on the same category:
  @netilon/differify x 1,045,377 ops/sec ±1.42% (93 runs sampled)

      deep-object-diff x 184,838 ops/sec ±2.55% (85 runs sampled)

      recursive-diff x 108,276 ops/sec ±1.93% (94 runs sampled)

      Fastest is @netilon/differify

## Whats new?

- Completely rewritten
- The new version 4.x is <font size="3"> **x2 faster**</font> than the older versions (version < 3.0.0)
  and is now one of the **Fastests deep object/array** comparators.
- Typescript support added.
- Support for **Node.js** and **Browsers** (it works on both)
- Just **7.6K (gzipped 2K)** weight (import)
- **No dependencies**
- **New features** were added! Now you can easily do more things with differify!
  - new config option added. Now, you can decide whether you prefer to compare arrays, either in an `ordered` or in an `unordered` way. Remember that, by default, you have an ordered comparison.
  - you can apply changes (merge) from `left to right` (applyRightChanges) or `right to left` (applyLeftChanges)
  - you can just `keep the differences between two entities` It's very useful indeed! (see more in the [Documentation](#id3) about the diffOnly option of `apply[Right|Left]Changes` methods).
  - you can filter the diff result of `compare()` method by an specific status (`ADDED`, `MODIFIED`, `DELETED`, `EQUAL`).

## Synopsis

Differify allows you to get the diff between two entities (objects diff, arrays diff, date diff, functions diff, number diff, etc) very easily, quickly and in a friendly way.

<img alt="GitHub Workflow Status" src="https://img.shields.io/github/workflow/status/netilon/differify/Differify CI"><img alt="GitHub package.json version" src="https://img.shields.io/github/package-json/v/netilon/differify">

![](badges/badge-functions.svg)
![](badges/badge-lines.svg)
![](badges/badge-statements.svg)

## Your contribution is appreciated (thanks!)

[![alt text](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif 'thanks for contribute!')](https://paypal.me/netilon)

---

**Index**

1. [Installation](#id1)
2. [How to use](#id2)
3. [Documentation](#id3)
4. [Configuration](#id4)
5. [Typescript](#id5)
6. [Examples](#id6)

---

## Installation<a name="id1"></a>

npm install @netilon/differify

## How to use it<a name="id2"></a>

Comparing things with differify is **very simple**!

### **> Compare objects**

![](assets/how-to-use-differify-object-example.png)

### **> Object diff output**

![](assets/differify-object-output.png)

### **> Easy access and use**

![](assets/basic-use.png)

### **> Compare arrays**

![](assets/how-to-use-differify-array-example.png)

### **> Array diff output**

![](assets/differify-array-output.png)

### **Simple Structure**

As you can see, there are two different kinds of structures that you can get from `compare` method call.

1. For objects and arrays **only**, you will get this structure:

   ![](assets/basic-structure-for-objects-arrays.png)

   - the `_` property contains the detailed diff information (it's an underscore to improve the readability in complex nested objects property accesses)
   - the `status` property contains the global status of the comparison ('EQUAL', 'MODIFIED', 'DELETED', 'ADDED')
   - the `changes` property is the total changes found when the comparison was performed.

2. For anything that `Object.prototype.toString.call()` does NOT return `[object Array]` or `[object Object]` (functions, dates, numbers, etc), you will get this structure:

   ![](assets/basic-structure-for-values.png)

   - the `original` property has the **original** value (left parameter in `compare` method).
   - the `current` property has the **current** value (right parameter in `compare` method).
   - the `status` property contains the current status of the comparison ('EQUAL', 'MODIFIED', 'DELETED', 'ADDED')
   - the `changes` property will be 1 or 0 depending if there was a change or not.

### **> Apply changes**

![](assets/apply-changes.png)

# Documentation<a name="id3"></a>

## Methods

**Method:**

_setConfig(*object*);_

**Description:** It sets the configuration options that will be applied when compare() method is called.

**Params:**

Configuration Object (see the Configuration section).

**Return:** void

---

**Method:**

_getConfig();_

**Description:** It returns a copy of the current configuration object.

**Return:** Object

---

**Method:**

_compare(*any*, *any*);_

**Description:** It returns the difference between two entities.

**Params:**

Both parameters indicate the entities to be compared.

**Return:** Object.

---

**Method:**

_applyRightChanges(diffResult, diffOnly);_

**Description:** It will apply the changes (merge both entities) and will keep the modified values **from the right**.

**Params:**

_diffResult_: Object - It is the Object returned by the `compare()` method call.

_diffOnly_: boolean - (default: false) It returns just the difference (only the !== `EQUAL` properties).

**Return:** Object.

---

**Method:**

_applyLeftChanges(diffResult, diffOnly);_

**Description:** It will apply the changes (merge both entities) and will keep the modified values **from the left**.

**Params:**

_diffResult_: Object - It is the Object returned by the `compare()` method call.

_diffOnly_: boolean - (default: false) It returns just the difference (only the !== `EQUAL` properties).

**Return:** Object.

---

**Method:**

_filterDiffByStatus(diffResult, status, extendedInformation);_

**Description:** It will return the changes that match the specified property status (second parameter) using the **DIFF_MODES.DIFF** in the configuration. **IMPORTANT:** If you use another diff mode (different than DIFF_MODES.DIFF), then you will get the corresponding value from original/current property (based on the status) without any filtering (you will get the raw value).

**Params:**

_diffResult_: Object - It is the Object returned by the `compare()` method call.

_status_: string - one of the following (`ADDED` || `MODIFIED` || `DELETED` || `EQUAL`).

_extendedInformation_: boolean - if true, it will add more detail about the elements to the given output. Defaults to false.

**Return:** any || null - it depends on the input type. If there is no status matching, then null will be returned.

---

## Configuration<a name="id4"></a>

You can pass a config to the setConfig() method to change the behavior and adjust it to fit your needs. If you prefer, you can set it once and use it everywhere or you can change it when you need it.

| key                    | value   | default   | description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
| ---------------------- | ------- | --------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| _mode.array_           | string  | DIFF      | **DIFF**: it will iterate over each element in the array A, and will compare each element against the element in the same index in the array B.<br><br>**REFERENCE**: just compare the references of each array.<br><br>**STRING**: only will check the array length and will do a toString comparison if necessary.                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| _mode.object_          | string  | DIFF      | **DIFF**: it will iterate over each property in the object A and will compare each value with the same property value in the object B.<br><br>**REFERENCE**: just compare the references of each object.<br><br>**STRING**: only do a toString comparison.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| _mode.function_        | string  | REFERENCE | **REFERENCE**: just compare the references of each function.<br><br>**STRING**: only do a toString comparison (useful to compare the function bodies).                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
| _compareArraysInOrder_ | boolean | true      | if it is `true`, it will compare each element in both arrays one by one in an ordered way, assuming that each element, in the same index in both arrays, should be the same element that can have changes or not ([see an example of this case](#id6)). If the option is set to `false`, then it will compare each element in both arrays and it will check if they are `EQUAL`, `ADDED` or `DELETED` without keeping in mind the appearence order (there won't be details about the `MODIFIED` status, it's only available if the option is set to true, since in that case, the order matters and we know that each element in the same index (but in different arrays), should be the same element that could have been changed or not) ([see an example of this case](#id7)). |

**Configuration example:**

    const Differify = require('@netilon/differify');

    differify = new Differify({ mode: { object: 'DIFF', array: 'DIFF' } });

    const diff = differify.compare(a, b);

if you dont specify any configuration, the default options are the following:

    {
        mode: {
        array: 'DIFF',
        object: 'DIFF',
        function: 'REFERENCE',
    }

## Typescript<a name="id5"></a>

To use the differify library with Typescript, you have to configure your tsconfig.json file and enable the following properties:

tsconfig.json

    {
    	"compilerOptions": {
    		"allowJs": true,
    		"esModuleInterop": true,
    	}
    }

then in your .ts file, you can import Differify this way:

    import Differify, { DIFF_MODES } from '@netilon/differify';

    // See the examples section.

## Examples<a name="id6"></a>

You have to know that the configuration you provide will change the behavior of the comparators and it will result in different outputs.

Just play around with it and use the configuration that fits your needs.

The following image, **just represents the idea** of what each option does, but **is not** the real implementation:

![](assets/comparators.png)

_Eg:_

### with the option **_DIFF_**:

    const testA = [1,2];
    const testB = [1,3];

you will get this output (note that there is a detail for each element in the array A and B):

    {
    	"_": [{
    		"original": 1,
    		"current": 1,
    		"status": "EQUAL",
    		"changes": 0
    	}, {
    		"original": 2,
    		"current": 3,
    		"status": "MODIFIED",
    		"changes": 1
    	}],
    	"status": "MODIFIED",
    	"changes": 1
    }

### with the option **_STRING_** or **_REFERENCE_**:

    const testA = [1,2];
    const testB = [1,3];

you will get this output (just a string comparison):

    {
    	"original": [1,2],
    	"current": [1,3],
    	"status": "MODIFIED",
    	// always will be 1 or 0 because there is no
    	// deep checking (use the DIFF option if you want more information)
    	"changes": 1
    }

### Array comparison example, keeping the order (compareArraysInOrder: true) <a name="id6"></a>

    const differify = new Differify({
    	compareArraysInOrder: true, //default value
    	mode: { object: 'DIFF', array: 'DIFF' },
    });

    const diff = differify.compare(['a', 'b'], ['a', 'z', 'b']);

    /*
    	OUTPUT

    	{
    		"_": [{
    			"original": "a",
    			"current": "a",
    			"status": "EQUAL",
    			"changes": 0
    		}, {
    			"original": "b",
    			"current": "z",
    			"status": "MODIFIED",
    			"changes": 1
    		}, {
    			"original": null,
    			"current": "b",
    			"status": "ADDED",
    			"changes": 1
    		}],
    		"status": "MODIFIED",
    		"changes": 2
    	}

    */

### Array comparison example, without having the order in mind (compareArraysInOrder: false).. <a name="id7"></a>

**NOTE**: In case you have Object elements inside the array and you are using compareArraysInOrder to false and array mode to true, then it will compare the objects as serialized strings to know if they are equal or not.

    const differify = new Differify({
    	compareArraysInOrder: false,
    	mode: { object: 'DIFF', array: 'DIFF' },
    });

    const diff = differify.compare(['a', 'b'], ['a', 'z', 'b']);

    /*
    	OUTPUT


    	{
    		"_": [{
    			"original": "a",
    			"current": "a",
    			"status": "EQUAL",
    			"changes": 0
    		}, {
    			"original": "b",
    			"current": "b",
    			"status": "EQUAL",
    			"changes": 0
    		}, {
    			"original": null,
    			"current": "z",
    			"status": "ADDED",
    			"changes": 1
    		}],
    		"status": "MODIFIED",
    		"changes": 1
    	}

    */

## Your contribution is appreciated (thanks!)

[![alt text](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif 'thanks for contribute!')](https://paypal.me/netilon)


CHANGELOG.md
- [2023-06-05 21:49:34 -0300] (8aac961) fixing vulnerabilities + updating dependencies
- [2021-06-02 23:38:40 -0300] (2e29448) Merge pull request #40 from netilon/hotfix_issue36
- [2021-06-02 23:01:37 -0300] (8ae3397) adding more test to cover issue36
- [2021-06-02 22:46:23 -0300] (fb9acba) updating README
- [2021-06-02 22:19:27 -0300] (d19a837) adding changelog
- [2021-06-02 22:13:39 -0300] (0fc2279) now the input data for the 'compare' method is included in its response (inside the 'original'/'current' properties). Before this update, if you used a diff mode different than DIFF_MODES.DIFF, then you got a null value for the '_' property, but now what you will get is not the '_' property but the properties 'original' and 'current' filled with the input data (first and second parameter). It will be usefull to use this output with methods like applyLeftChanges, applyRightChanges and filterDiffByStatus. This only apply when you use DIFF_MODES.REFERENCE OR DIFF_MODES.STRING to make them compatible with the mentioned methods.If you use the DIFF_MODES.DIFF in the config, the output will be the same.
- [2021-03-21 13:16:30 -0300] (c981db5) Merge pull request #31 from netilon/feature_issue30
- [2021-03-21 13:10:19 -0300] (2749631) removing redundant information in the extended filtered output
- [2021-03-21 12:43:30 -0300] (007b871) updating documentation
- [2021-03-21 12:27:19 -0300] (5f1da8a) adding feature for filterDiffByStatus method to be able to get extended information for each element in the output (issue30) + adding tests to address this new feature + upgrading package version

src/differify2.0.js
/*!
 * Differify v2.0.0
 * http://netilon.com/
 *
 * Copyright 2018 Netilon (Fabian Orue)
 * Released under the MIT license
 *
 * Date: 2018-06-02 20:35 GMT-0300 (ART)
 */

'use strict';





function isArray(value) {
    if (Object.prototype.toString.call(value) === '[object Array]') {
        return true;
    }
    return false;
}

function isObject(value) {
    if (!isArray(value) && (typeof value) === 'object') {
        return true;
    }
    return false;
}

function isString(value) {
    if (typeof value === 'string') {
        return true;
    }
    return false;
}

function isFunction(value) {
    if (typeof value === 'function') {
        return true;
    }
    return false;
}

function isDate(value) {
    if (Object.prototype.toString.call(value) === '[object Date]') {
        return true;
    }

    return false;
}

function isDefined(value) {
    if (value !== undefined && value !== null) {
        return true;
    }

    return false;
}

function ifNotDefinedGetDefault(object, property, defaultValue) {
    return object !== undefined && isDefined(object[property]) ? object[property] : defaultValue;
}


function clone(obj) {
    if (null == obj || "object" != (typeof obj)) {
        return obj;
    }
    var copy = obj.constructor();
    for (var attr in obj) {
        if (obj.hasOwnProperty(attr))
            copy[attr] = obj[attr];
    }
    return copy;
}

var returnTypes = {
    ARRAY: 'array',
    JSON_OBJECT: 'json'
};

function Configuration(_config) {
    this.deep = ifNotDefinedGetDefault(_config, 'deep', 3);
    this.scan = ifNotDefinedGetDefault(_config, 'scan', {arrays: true});
    this.returnType = ifNotDefinedGetDefault(_config, 'returnType', returnTypes.JSON_OBJECT);
}

var Differify = function () {
    var config = new Configuration();
    var deepDataCounter = 0;

    var valueStatus = {
        'ADDED': 'added',
        'DELETED': 'deleted',
        'MODIFIED': 'modified'
    };

    var dataTypes = {
        UNDEFINED: 0,
        OBJECT: 1,
        ARRAY: 2,
        STRING: 3,
        NUMBER: 4,
        FUNCTION: 5,
        DATE: 6
    };

    this.setConfig = function (_config) {
        config = new Configuration(_config);
    };


    this.objectDiff = function(a, b, path) {
        var ret = new Response(config.returnType);
        
        if (deepDataCounter <= 0) {
            return ret.getData();
        }
        
        path = getParentPath(path);

        var bCopy = clone(b);
        var typeA = null;
        var typeB = null;

        for (var property in a) {
            if (a.hasOwnProperty(property)) {
                if (b.hasOwnProperty(property) && b[property] !== undefined) {

                    delete bCopy[property];

                    typeA = getTypeOf(a[property]);
                    typeB = getTypeOf(b[property]);

                    if (typeA !== typeB) {
                        //if data types are different.
                        ret.push(propertyData(path, property, typeA === dataTypes.FUNCTION ? a[property].toString() : a[property], typeB === dataTypes.FUNCTION ? b[property].toString() : b[property], valueStatus.MODIFIED), property);
                        
                    } else if (typeA === dataTypes.UNDEFINED && typeB === dataTypes.UNDEFINED) {
                        continue;
                        
                    } else if (typeA === dataTypes.DATE && typeB === dataTypes.DATE) {
                        
                        if (areDateObjectsDifferent(a[property], b[property])) {
                            ret.push(propertyData(path, property, a[property], b[property], valueStatus.MODIFIED), property);
                        }
                        continue;
                        
                    } else if (typeA === dataTypes.ARRAY && typeB === dataTypes.ARRAY) {

                        var arrayCompareResult = this.arrayDiff(a[property], b[property], config.scan.arrays);
                        if (arrayCompareResult.length > 0) {
                            ret.push(propertyData(path, property, a[property], arrayCompareResult, valueStatus.MODIFIED), property);
                        }
                        continue;
                        
                    } else if (typeA === dataTypes.OBJECT && typeB === dataTypes.OBJECT) {
                        
                        --deepDataCounter;
                        ret.concat(this.objectDiff(a[property], b[property], path + '.' + property), property);
                        continue;
                        
                    } else if (typeA === dataTypes.FUNCTION && typeB === dataTypes.FUNCTION) {
                        
                        if (areFunctionsDifferent(a[property], b[property])) {
                            ret.push(propertyData(path, property, a[property].toString(), b[property].toString(), valueStatus.MODIFIED), property);
                        }
                    } else if (a[property] !== b[property]) {
                        //native type evaluation.
                        ret.push(propertyData(path, property, a[property], b[property], valueStatus.MODIFIED), property);
                    }
                } else {
                    ret.push(propertyData(path, property, a[property], null, valueStatus.DELETED), property);
                }
            }
        }
        property = null;
        for (property in bCopy) {
            if (b.hasOwnProperty(property)) {
                ret.push(propertyData(path, property, null, bCopy[property], valueStatus.ADDED), property);
            }
        }

        return ret.getData();
    }

    this.arrayDiff = function (arrayA, arrayB, compareEachElement) {
        compareEachElement = compareEachElement || false;
        var ret = [];

        if (!compareEachElement) {

            if (arrayA.length !== arrayB.length) {
                ret.push(valueData(arrayA, arrayB, valueStatus.MODIFIED));
            } else {
                var a = arrayA.toString();
                var b = arrayB.toString();
                if (a !== b) {
                    ret.push(valueData(arrayA, arrayB, valueStatus.MODIFIED));
                }
            }

            return ret;
        }

        var AisMajorThanB = arrayA.length - arrayB.length;
        var loop = AisMajorThanB < 0 ? arrayA.length : arrayB.length;

        for (var i = 0; i < loop; ++i) {

            if (isObject(arrayA[i]) && isObject(arrayB[i])) {
                var _a = JSON.stringify(arrayA[i]);
                var _b = JSON.stringify(arrayB[i]);
                if (_a !== _b) {
                    ret.push(valueData(_a, _b, valueStatus.MODIFIED));
                }
            } else if (arrayA[i] !== arrayB[i]) {
                ret.push(valueData(arrayA[i], arrayB[i], valueStatus.MODIFIED));
            }
        }

        if (AisMajorThanB > 0) {
            for (i = i; i < arrayA.length; ++i) {
                ret.push(valueData(arrayA[i], null, valueStatus.DELETED));
            }
        } else if (AisMajorThanB < 0) {
            for (i = i; i < arrayB.length; ++i) {
                ret.push(valueData(null, arrayB[i], valueStatus.ADDED));
            }
        }
        return ret;
    }

    function areDateObjectsDifferent(a, b) {
        if (a.getTime() !== b.getTime()) {
            return true;
        }
        return false;
    }

    function areFunctionsDifferent(a, b) {
        a = a.toString();
        b = b.toString();
        if (a !== b) {
            return true;
        }
        return false;
    }
    

    this.getDiff = function (a, b) {
        
        var ret = new Response(config.returnType);
        deepDataCounter = config.deep;
        

        var typeA = getTypeOf(a);
        var typeB = getTypeOf(b);

        var path = '$root';
        
        if (typeA !== typeB) {
            ret.push(propertyData(path, null, typeA === types.FUNCTION ? a.toString() : a, typeB === types.FUNCTION ? b.toString() : b, valueStatus.MODIFIED), path);
            return ret;
        } else if (typeA === dataTypes.OBJECT && typeB === dataTypes.OBJECT) {
            ret.concat(this.objectDiff(a,b,path));
        } else if (typeA === dataTypes.DATE && typeB === dataTypes.DATE) {
            if (areDateObjectsDifferent(a, b)) {
                ret.push(propertyData(path, property, a, b, valueStatus.MODIFIED), path);
            }
        } else if (typeA === dataTypes.ARRAY && typeB === dataTypes.ARRAY) {
            var arrayCompareResult = this.arrayDiff(a, b, config.scan.arrays);
            if (arrayCompareResult.length > 0) {
                ret.push(propertyData(path, null, a, arrayCompareResult, valueStatus.MODIFIED), path);
            }
        } else {
            if ((a = a.toString()) !== (b = b.toString())) {
                ret.push(propertyData(path, null, a, b, valueStatus.MODIFIED), path);
            }
        }
        
        return ret.getData();
    }


    //Helpers

    function getParentPath(parent) {
        return ((typeof parent === 'string' && parent.trim() !== '') ? parent : '$root');
    }

    function getTypeOf(value) {

        if (!isDefined(value)) {
            return dataTypes.UNDEFINED;
        } else if (isString(value)) {
            return dataTypes.STRING;
        } else if (isArray(value)) {
            return dataTypes.ARRAY;
        } else if (isDate(value)) {
            return dataTypes.DATE;
        } else if (isObject(value)) {
            return dataTypes.OBJECT;
        } else if (isFunction(value)) {
            return dataTypes.FUNCTION;
        }

        return dataTypes.NUMBER;
    }
    
    function Response(format){
        var data = null;
        this.push = null;
        this.clear = null;
        this.concat = null;
        
        if(format === returnTypes.JSON_OBJECT){
            data = {};
            this.push = function push(_data, propertyName){
                data[propertyName] = _data;
            };
            this.clear = function clear(){
                data = {};
            };
            
            this.concat = function concat(response, property){
                if(property===undefined){
                    data = Object.assign(data, response);
                }else{
                    data[property] = response;
                }
            };
        }else{
            
            data = [];
            this.push = function push(_data, propertyName){
                data.push(_data);
            };
            this.clear = function clear(){
                data = [];
            };
            
            this.concat = function concat(response, property){
                data = data.concat(response);
            };
        }
        
        this.clear();
        
        this.getData = function(){
            return data;
        }
    }

    function valueData(valueA, valueB, status) {
        return {
            original: valueA,
            diff: valueB,
            status: status
        };
    };

    function propertyData(path, property, valueA, valueB, status) {
        return {
            path: path !== '' && property !== null ? path + '.' + property : path,
            property: property,
            value: valueData(valueA, valueB, status)
        };
    };
}



module.exports = new Differify();